clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -0,0 +1,229 @@
+            public void write(ByteBuffer buffer, Object o) {
+                Assignment assignment = (Assignment) o;
+                assignment.struct.writeTo(buffer);
+            }
+
+            @Override
+            public Object read(ByteBuffer buffer) {
+                Struct struct = (Struct) ASSIGNMENT_V0.read(buffer);
+                return new Assignment(struct);
+            }
+
+            @Override
+            public int sizeOf(Object o) {
+                Assignment assignment = (Assignment) o;
+                return assignment.struct.sizeOf();
+            }
+        };
+    }
+
+    public static class Assignment {
+        private final Struct struct;
+        private final short version;
+        private final List<TopicPartition> partitions;
+
+        private Map<String, List<Integer>> asMap(Collection<TopicPartition> partitions) {
+            Map<String, List<Integer>> partitionMap = new HashMap<>();
+            for (TopicPartition partition : partitions) {
+                String topic = partition.topic();
+                List<Integer> topicPartitions = partitionMap.get(topic);
+                if (topicPartitions == null) {
+                    topicPartitions = new ArrayList<>();
+                    partitionMap.put(topic, topicPartitions);
+                }
+                topicPartitions.add(partition.partition());
+            }
+            return partitionMap;
+        }
+
+        public Assignment(List<TopicPartition> partitions) {
+            this.struct = new Struct(ASSIGNMENT_V0);
+
+            List<Struct> topicAssignments = new ArrayList<>();
+            for (Map.Entry<String, List<Integer>> topicEntry : asMap(partitions).entrySet()) {
+                Struct topicAssignment = new Struct(TOPIC_ASSIGNMENT_V0);
+                topicAssignment.set(TOPIC_KEY_NAME, topicEntry.getKey());
+                topicAssignment.set(PARTITIONS_KEY_NAME, topicEntry.getValue().toArray());
+                topicAssignments.add(topicAssignment);
+            }
+            this.struct.set(ASSIGNMENT_KEY_NAME, topicAssignments.toArray());
+            this.partitions = partitions;
+
+            this.version = CONSUMER_PROTOCOL_V0;
+            this.struct.set(VERSION_KEY_NAME, CONSUMER_PROTOCOL_V0);
+        }
+
+        public Assignment(Struct struct) {
+            this.struct = struct;
+            this.partitions = new ArrayList<>();
+            for (Object structObj : struct.getArray(ASSIGNMENT_KEY_NAME)) {
+                Struct assignment = (Struct) structObj;
+                String topic = assignment.getString(TOPIC_KEY_NAME);
+                for (Object partitionObj : assignment.getArray(PARTITIONS_KEY_NAME)) {
+                    Integer partition = (Integer) partitionObj;
+                    this.partitions.add(new TopicPartition(topic, partition));
+                }
+            }
+            this.version = struct.getShort(VERSION_KEY_NAME);
+        }
+
+        public short version() {
```

Comment:
The leader provides compatible versions of assignments given the respective versions specified in member subscriptions. For this patch, I have assumed full forwards and backwards compatibility of the embedded group protocol, which allows any member to be elected leader during an upgrade scenario and still perform correctly. There are two cases to consider:

1. The leader is on the old version: In this case, forwards compatibility allows the leader to go ahead and parse subscriptions from all members (even those on newer versions) and generate assignments corresponding to its older version. Backwards compatibility ensures that the newer members will be able to parse the assignments with older versions. For example, if the leader is on version 0 and a follower is on version 1, the leader will parse the version 1 subscription as a version 0 subscription and provide it a version 0 assignment.
2. The leader is on the new version: In this case, the leader can parse both new and old subscription versions. Depending on the protocol, it can choose to send assignments to members corresponding to their respective subscriptions, or it can choose the oldest version and send assignments based on it. Using a similar example as above, if the leader is on version 1 and the follower is on version 0, then the leader can parse the version 0 subscription and send a version 0 assignment.

Obviously this strict compatibility model might be challenging to implement in practice, depending in particular on the protocol format (it would be relatively straightforward if we chose JSON, for example). Alternatively, if we exposed version information in the JoinGroup protocol, then the coordinator would be able choose the member with the highest version, which would let us weaken the compatibility assumption. The tradeoff is that we'd need to complicate the protocol a bit to carry version information from each member along with their respective subscriptions and assignments. 

This gets even trickier if you allow assignment strategies to include their own embedded data format, so I was hoping to get a simple approach in place first before having a more complete discussion. Perhaps for now I can just add some documentation on the GroupProtocol class to clarify the current compatibility assumptions?